runtime/commonMain/src/kotlinx/serialization/json/JsonElementBuilders.kt

```diff
@@ -1,101 +1,248 @@
+public class JsonObjectBuilder internal constructor() {
 
-    internal val content: MutableList<JsonElement> = mutableListOf()
+    private val content: MutableMap<String, JsonElement> = linkedMapOf()
 
     /**
-     * Adds [this] value to the current [JsonArray] as [JsonPrimitive].
+     * Add the [JSON][JsonObject] produced by the [builderAction] function
+     * to a resulting json object using the given [key].
      */
-    public operator fun String?.unaryPlus() {
-        content.add(JsonPrimitive(this))
+    public fun addJson(key: String, builderAction: JsonObjectBuilder.() -> Unit) {
+        content[key] = buildJson(builderAction)
     }
 
     /**
-     * Adds [this] value to the current [JsonArray] as [JsonPrimitive].
+     * Add the [JSON array][JsonArray] produced by the [builderAction] function
+     * to a resulting json object using the given [key].
      */
-    public operator fun Number?.unaryPlus() {
-        content.add(JsonPrimitive(this))
+    public fun addArray(key: String, builderAction: JsonArrayBuilder.() -> Unit) {
+        content[key] = buildJsonArray(builderAction)
     }
 
     /**
-     * Adds [this] value to the current [JsonArray] as [JsonPrimitive].
+     * Add the given JSON [element] to a resulting JSON object using the given [key].
      */
-    public operator fun Boolean?.unaryPlus() {
-        content.add(JsonPrimitive(this))
+    public fun add(key: String, element: JsonElement) {
+        content[key] = element
     }
 
     /**
-     * Adds [this] value to the current [JsonArray].
+     * Add the given boolean [value] to a resulting JSON object using the given [key].
      */
-    public operator fun JsonElement.unaryPlus() {
-        this@JsonArrayBuilder.content.add(this)
+    public fun add(key: String, value: Boolean?) {
+        content[key] = JsonPrimitive(value)
     }
-}
 
-/**
- * DSL builder for a [JsonObject]. To create an instance of builder, use [json].
- */
-public class JsonObjectBuilder internal constructor() {
+    /**
+     * Add the given numeric [value] to a resulting JSON object using the given [key].
+     */
+    public fun add(key: String, value: Number?) {
+        content[key] = JsonPrimitive(value)
+    }
+
+    /**
+     * Add the given string [value] to a resulting JSON object using the given [key].
+     */
+    public fun add(key: String, value: String?) {
+        content[key] = JsonPrimitive(value)
+    }
 
-    internal val content: MutableMap<String, JsonElement> = linkedMapOf()
+    internal fun build(): JsonObject = JsonObject(content)
 
     /**
      * Adds given [value] to the current [JsonObject] with [this] as a key.
      */
+    @Deprecated(message = infixToDeprecated, replaceWith = ReplaceWith("add(this, value)"))
     public infix fun String.to(value: JsonElement) {
-        require(content[this] == null) {"Key $this is already registered in builder"}
+        require(content[this] == null) { "Key $this is already registered in builder" }
```

Comment:
This shows the difference between old and new behavior: old prohibits overwriting, new allows it. I'm not sure which one is better, but probably changing it without notice is not good either.

Reply:
I am fine with both, decided not to check it to be consistent with `buildMap`